### DISCUSSION TITLE
相机的操作方式是否需要参考一下主流的三维建模软件？

### DISCUSSION THREAD

**Discussion by Mostro-Complexity** [Acmen-Team/Epoch#6] [General]
目前相机视角的切换有一些蹩脚，可否调整一下？😉

**Comment by Holy-YxY** [comment]
> 目前相机视角的切换有一些蹩脚，可否调整一下？😉

你可以简单描述下蹩脚的主要来源和感受，另外大家可以先就目前的方式做个讨论，讨论的差不多了我们就切入到实际的开发中去。

我先就目前的方式以及相关存在问题做简要介绍（目前也比较简单，就是相关IO事件做变换计算）

## 目前的操作方式汇总
- W、S：相机前后的平移变换
- A、D：相机左右的平移变换
- Q、E：相机上下的平移变换
- 鼠标滚轮：相机前后的平移变化
- 鼠标右键按住然后拖动：相机以自身为圆心的旋转变换
## 目前存在的一些问题
- IO事件响应的冲突（例如鼠标焦点在其他window上时，滚动鼠标相机也会响应IO事件做出相应变换）
- window上IO事件的响应目前是以点击选中后再做响应，这样不够灵活。（例如某些window上的组件可能希望只需要鼠标悬停在上面就可以对其操作）

**Reply by Mostro-Complexity** [reply]
我个人更偏爱blender的操作方式：https://www.bilibili.com/read/cv10547329/
其中涉及到了鼠标中键、shift键、alt键和一个额外辅助键位的组合。其设计我认为值得参考。
至于你提到的问题中的第二点，blender也存在类似的问题。可能要在UI框架中的事件触发机制上做研究。

除此之外，还需要有编辑器上的图形拾取；
双击在右边的Scene Hierarchy中双击对象，应该在编辑器上能够定位到对应的图形的位置。

**Reply by Holy-YxY** [reply]
看了下链接里演示的效果，表示有的非常认可。
但是blender因为有建模功能需求，所以追求更精细化的控制。Epoch的设计初衷是渲染引擎，并不想带来太多的建模方面的操作，针对这些我觉得可以在现有的效果和blender之间做个取中，即追加blender一些我们觉得有必要的操作的同时，优化或直接改变现在已有的一些操作。最后对于你提到的拾取，也就是双击定位，这个非常有必要，优先度我认为是这个讨论里最高的。

### 拾取定位功能作为第一优先级

### 下面我总结下我觉得第一个版本（后续可以继续追加完善）最终应该具有的操作合集以及对应的键位组合
- 放大缩小视图 -- 鼠标中键滚动
- 平移视图  -- Shift+鼠标右键按住拖动
- 旋转视图 -- 鼠标右键按住旋转
- 隔离模式 -- Shift+Alt(前提是当前有被选中的对象)
- 线框模式 -- Shift+Z(前提是当前有被选中的对象)
PS：现有的键盘操作依旧保留

**Reply by Holy-YxY** [reply]
关于图形拾取，也就是双击对象定位到图形位置这个操作，需要讨论下具体怎么实现，以及一些实现细节。
我目前能想到一种简单方法就是移动到以当前相机视域方向目标对象在屏幕中居中的位置，关于目标对象居中的判断可以在NDC空间下做判断。

**Reply by Mostro-Complexity** [reply]
> 关于图形拾取，也就是双击对象定位到图形位置这个操作，需要讨论下具体怎么实现，以及一些实现细节。 我目前能想到一种简单方法就是移动到以当前相机视域方向目标对象在屏幕中居中的位置，关于目标对象居中的判断可以在NDC空间下做判断。

这样的思路可以的，可以着手开始实现了。但是有一点，图形拾取不仅要能够从侧边的导航栏中选中，在编辑场景中也要可以选中。这种操作在侧边栏对象较多时会很有用。下图以Unity为例。
![image](https://user-images.githubusercontent.com/18632298/147079057-178c8995-b33a-4324-90c2-8a155278ef19.png)

**Reply by Holy-YxY** [reply]
针对提到的编辑场景中选中这个是在原先的计划之内的，感兴趣的话可以看Cherno这个视频了解相关原理：[Mouse Picking // Game Engine series](https://www.youtube.com/watch?v=GKERbQ7Sqyg)

**Reply by Mostro-Complexity** [reply]
如果是在计划之内，那就没有问题了。
接下来我先尝试着编写一些项目的开发者文档，将我们设想中的一些功能先以文本的形式记录下来，这样也方便其他加入的人着手工作。
